pkg/query/statistics.go

```diff
@@ -0,0 +1,210 @@
+		}, nil
+	case *IntersectionArrow:
+		ls, err := s.Cost(it.left)
+		if err != nil {
+			return Estimate{}, err
+		}
+		rs, err := s.Cost(it.right)
+		if err != nil {
+			return Estimate{}, err
+		}
+		// IntersectionArrow also does IterSubjects on left, then Check on right,
+		// but only yields results when ALL subjects satisfy (more selective)
+		selectivity := ls.CheckSelectivity * rs.CheckSelectivity * s.IntersectionArrowReduction // Additional reduction for "all" semantics
+		return Estimate{
+			Cardinality:       int(float64(ls.Cardinality*rs.Cardinality) * selectivity),
+			CheckCost:         ls.IterSubjectsCost + (ls.Cardinality * rs.CheckCost),
+			CheckSelectivity:  selectivity,
+			IterResourcesCost: ls.IterResourcesCost + (ls.Cardinality * rs.IterResourcesCost),
+			IterSubjectsCost:  ls.IterSubjectsCost + (ls.Cardinality * rs.IterSubjectsCost),
+		}, nil
+	case *Union:
+		if len(it.subIts) == 0 {
+			return Estimate{}, errors.New("StaticStatistics: union with no subiterators")
+		}
+
+		// Union sums up all subiterator costs
+		result := Estimate{}
+		for _, subIt := range it.subIts {
+			est, err := s.Cost(subIt)
+			if err != nil {
+				return Estimate{}, err
+			}
+			result.Cardinality += est.Cardinality
+			result.CheckCost += est.CheckCost
+			result.IterResourcesCost += est.IterResourcesCost
+			result.IterSubjectsCost += est.IterSubjectsCost
+			if est.CheckSelectivity > result.CheckSelectivity {
+				result.CheckSelectivity = est.CheckSelectivity
+			}
+		}
+		return result, nil
+	case *Intersection:
+		if len(it.subIts) == 0 {
+			return Estimate{}, errors.New("StaticStatistics: intersection with no subiterators")
+		}
+
+		// Intersection: costs add up, but cardinality decreases with each iterator
+		result := Estimate{CheckSelectivity: 1}
+		for idx, subIt := range it.subIts {
+			est, err := s.Cost(subIt)
+			if err != nil {
+				return Estimate{}, err
+			}
+
+			if idx == 0 {
+				result.Cardinality = est.Cardinality
+			} else {
+				// Each subsequent intersection filters down the results
+				result.Cardinality = int(float64(result.Cardinality) * est.CheckSelectivity)
+			}
+
+			result.CheckCost += est.CheckCost
+			result.IterResourcesCost += est.IterResourcesCost
+			result.IterSubjectsCost += est.IterSubjectsCost
+			result.CheckSelectivity *= est.CheckSelectivity
+		}
+		return result, nil
+	case *Exclusion:
+		// Exclusion: A - B
+		// We need to check both sides, but the result is filtered
+		mainEst, err := s.Cost(it.mainSet)
+		if err != nil {
+			return Estimate{}, err
+		}
+		excludedEst, err := s.Cost(it.excluded)
+		if err != nil {
+			return Estimate{}, err
+		}
+
+		// Cardinality: main minus the excluded portion
+		// Approximate as: main.Cardinality * (1 - excluded.CheckSelectivity)
+		exclusionFactor := 1.0 - excludedEst.CheckSelectivity
+		if exclusionFactor < 0 {
+			exclusionFactor = 0
+		}
+
+		return Estimate{
+			Cardinality:       int(float64(mainEst.Cardinality) * exclusionFactor),
```

Comment:
similar to my other comment:  why is this not the sum of `main` and `excluded`? you are mixing counts with probabilities which doesn't make sense to me